python/cuml/ts/arima.pyx

```diff
@@ -0,0 +1,814 @@
+    # best_model =
+
+    for p in range(0, max_p):
+        arparams = np.zeros(p)
+        for q in range(0, max_q):
+            maparams = np.zeros(q)
+
+            # skip 0,0 case
+            if p == 0 and q == 0:
+                continue
+
+            mu0, ar0, ma0 = estimate_x0((p, d, q), 2, y_b)
+            
+            b_model = fit(y_b, (p, d, q), mu0, ar0, ma0)
+
+            if method == "aic":
+                ic = b_model.aic
+            elif method == "bic":
+                ic = b_model.bic
+            else:
+                raise NotImplementedError("Method '{}' not supported".format(method))
+
+            for (i, ic_i) in enumerate(ic):
+                if ic_i < best_ic[i]:
+                    best_model.order[i] = (p, d, q)
+                    best_model.mu[i] = b_model.mu[i]
+
+                    if p > 0:
+                        best_model.ar_params[i] = b_model.ar_params[i]
+                    else:
+                        best_model.ar_params[i] = []
+                    if q > 0:
+                        best_model.ma_params[i] = b_model.ma_params[i]
+                    else:
+                        best_model.ma_params[i] = []
+
+                    best_ic[i] = ic_i
+
+    return (best_model, best_ic)
+
+
+def unpack(p, d, q, nb, np.ndarray[double, ndim=1] x):
+    """Unpack linearized parameters into mu, ar, and ma batched-groupings"""
+    pynvtx_range_push("unpack(x) -> (ar,ma,mu)")
+    num_parameters = d + p + q
+    mu = np.zeros(nb)
+    ar = []
+    ma = []
+    for i in range(nb):
+        xi = x[i*num_parameters:(i+1)*num_parameters]
+        if d > 0:
+            mu[i] = xi[0]
+        if p > 0:
+            ar.append(xi[d:(d+p)])
+        ma.append(xi[d+p:])
+
+    pynvtx_range_pop()
+    return (mu, ar, ma)
+
+
+def pack(p, d, q, nb, mu, ar, ma):
+    """Pack mu, ar, and ma batched-groupings into a linearized vector `x`"""
+    pynvtx_range_push("pack(ar,ma,mu) -> x")
+    num_parameters = d + p + q
+    x = np.zeros(num_parameters*nb)
+    for i in range(nb):
+        xi = np.zeros(num_parameters)
+        if d > 0:
+            xi[0] = mu[i]
+
+        for j in range(p):
+            xi[j+d] = ar[i][j]
+        for j in range(q):
+            xi[j+p+d] = ma[i][j]
+        # xi = np.array([mu[i]])
+        # xi = np.r_[xi, ar[i]]
+        # xi = np.r_[xi, ma[i]]
+        x[i*num_parameters:(i+1)*num_parameters] = xi
+
+    pynvtx_range_pop()
+    return x
+
+
+def _batched_transform(p, d, q, nb, np.ndarray[double] x, isInv, handle=None):
+    cdef vector[double] vec_ar
+    cdef vector[double] vec_ma
+    cdef vector[double] vec_Tar
+    cdef vector[double] vec_Tma
+
+    pynvtx_range_push("batched_transform")
+    
+    if handle is None:
+        handle = cuml.common.handle.Handle()
+    cdef cumlHandle* handle_ = <cumlHandle*><size_t>handle.getHandle()
+    cdef np.ndarray[double] Tx = np.zeros(nb*(d+p+q))
```

Comment:
We probably don't want to use the numpy cython approach, just the normal python one that I mentioned above. Applies for all of the methods below in this file

Reply:
The `_batched_transform()` method is a bit special because it is needed for the CPU-only optimizer and thus requires CPU arrays. What is the "normal python approach" to allocate an array and get its pointer for CPU memory?